src/compiler/utilities.ts

```diff
@@ -2457,6 +2485,80 @@ namespace ts {
         declarationFilePath: string;
     }
 
+    export function getSourceFilesToEmit(host: EmitHost, targetSourceFile?: SourceFile) {
+        const options = host.getCompilerOptions();
+        if (options.outFile || options.out) {
+            const moduleKind = getEmitModuleKind(options);
+            const moduleEmitEnabled = moduleKind === ModuleKind.AMD || moduleKind === ModuleKind.System;
+            const sourceFiles = host.getSourceFiles();
+            // Can emit only sources that are not declaration file and are either non module code or module with --module or --target es6 specified
+            return filter(sourceFiles, moduleEmitEnabled ? isNonDeclarationFile : isBundleEmitNonExternalModule);
+        }
+        else {
+            const sourceFiles = targetSourceFile === undefined ? host.getSourceFiles() : [targetSourceFile];
+            return filter(sourceFiles, isNonDeclarationFile);
+        }
+    }
+
+    function isNonDeclarationFile(sourceFile: SourceFile) {
+        return !isDeclarationFile(sourceFile);
+    }
+
+    function isBundleEmitNonExternalModule(sourceFile: SourceFile) {
+        return !isDeclarationFile(sourceFile) && !isExternalModule(sourceFile);
+    }
+
+    export function forEachEmitFile(host: EmitHost, sourceFiles: SourceFile[],

```

Comment:
It will be great if you could comment the difference between `forEachEmitFile` `forEachExpectedEmitFile`. they are quite similar at a glace